tests/components/nest/sensor_sdm_test.py

```diff
@@ -0,0 +1,299 @@
+                        "ambientTemperatureCelsius": 25.1,
+                    },
+                    "sdm.devices.traits.Humidity": {
+                        "ambientHumidityPercent": 35.0,
+                    },
+                },
+            },
+            auth=None,
+        )
+    }
+    await setup_sensor(hass, devices)
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is not None
+    assert temperature.state == "25.1"
+
+    humidity = hass.states.get("sensor.my_sensor_humidity")
+    assert humidity is not None
+    assert humidity.state == "35.0"
+
+    registry = await hass.helpers.entity_registry.async_get_registry()
+    entry = registry.async_get("sensor.my_sensor_temperature")
+    assert entry.unique_id == "some-device-id-temperature"
+    assert entry.original_name == "My Sensor Temperature"
+    assert entry.domain == "sensor"
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    device = device_registry.async_get(entry.device_id)
+    assert device.name == "My Sensor"
+    assert device.model == "Thermostat"
+    assert device.identifiers == {("nest", "some-device-id")}
+
+
+async def test_no_devices(hass):
+    """Test no devices returned by the api."""
+    await setup_sensor(hass)
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is None
+
+    humidity = hass.states.get("sensor.my_sensor_humidity")
+    assert humidity is None
+
+
+async def test_device_no_sensor_traits(hass):
+    """Test a device with applicable sensor traits."""
+    devices = {
+        "some-device-id": Device.MakeDevice(
+            {
+                "name": "some-device-id",
+                "type": THERMOSTAT_TYPE,
+                "traits": {},
+            },
+            auth=None,
+        )
+    }
+    await setup_sensor(hass, devices)
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is None
+
+    humidity = hass.states.get("sensor.my_sensor_humidity")
+    assert humidity is None
+
+
+async def test_device_name_from_structure(hass):
+    """Test a device without a custom name, inferring name from structure."""
+    devices = {
+        "some-device-id": Device.MakeDevice(
+            {
+                "name": "some-device-id",
+                "type": THERMOSTAT_TYPE,
+                "traits": {
+                    "sdm.devices.traits.Temperature": {
+                        "ambientTemperatureCelsius": 25.2,
+                    },
+                },
+                "parentRelations": [
+                    {"parent": "some-structure-id", "displayName": "Some Room"}
+                ],
+            },
+            auth=None,
+        )
+    }
+    await setup_sensor(hass, devices)
+
+    temperature = hass.states.get("sensor.some_room_temperature")
+    assert temperature is not None
+    assert temperature.state == "25.2"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)